src/main/java/com/google/research/bleth/services/DatabaseService.java

```diff
@@ -0,0 +1,223 @@
+    public String readRealBoardState(String simulationId, int round) {
+        return readBoardState(simulationId, round, "RealBoard");
+    }
+
+    /**
+     * Read from EstimatedBoard and construct a Json string encoding the board state.
+     * @param simulationId is the simulation id to be retrieved.
+     * @param round is the round to be retrieved.
+     * @return a Json string encoding the board state, uses for visualization,
+     * or null if provided round exceeds simulation's maximum number of rounds.
+     */
+    public String readEstimatedBoardState(String simulationId, int round) {
+        return readBoardState(simulationId, round, "EstimatedBoard");
+    }
+
+    private String readBoardState(String simulationId, int round, String entityKind) {
+        // throw an exception if provided round exceeds simulation's maximum number of rounds.
+        if (!isRoundExistsInSimulation(simulationId, round)) {
+            throw new ExceedingRoundException("Provided round " + round +
+                    " exceeds maximum number of rounds of simulation " + simulationId);
+        }
+
+        // Set simple predicates.
+        Query.FilterPredicate filterBySimulationId =
+                new Query.FilterPredicate("simulationId", Query.FilterOperator.EQUAL, simulationId);
+        Query.FilterPredicate filterByRound =
+                new Query.FilterPredicate("round", Query.FilterOperator.EQUAL, round);
+
+        // Compose simple predicates.
+        Query.CompositeFilter composedFilter = Query.CompositeFilterOperator.and(filterBySimulationId, filterByRound);
+
+        // Create query and return result.
+        Query boardStateBySimulationIdAndRoundQuery = new Query(entityKind).setFilter(composedFilter);
+        PreparedQuery boardStateBySimulationIdAndRoundPreparedQuery = datastore.prepare(boardStateBySimulationIdAndRoundQuery);
+
+        // Return encoded board state.
+        return toJsonTable(boardStateBySimulationIdAndRoundPreparedQuery.asIterable(), simulationId);
+    }
+
+    private String toJsonTable(Iterable<Entity> boardStateEntities, String simulationId) {
+        // Retrieve simulations' dimensions.
+        Entity simulationEntity = retrieveSimulationEntity(simulationId);
+        int rowNum = ((Long) simulationEntity.getProperty("rowNum")).intValue();
+        int colNum = ((Long) simulationEntity.getProperty("colNum")).intValue();
+
+        // Create an empty table representing the board state.
+        ArrayTable<Integer, Integer, ArrayList<String>> boardState = createEmptyTable(rowNum, colNum);
+
+        // Fill board state with agents' string representations according to query result.
+        for (Entity entity : boardStateEntities) {
+            int row = ((Long) entity.getProperty("rowNum")).intValue();
+            int col = ((Long) entity.getProperty("colNum")).intValue();
+            String type = (String) entity.getProperty("agentType");
+            Long id = (Long) entity.getProperty("agentId");
```

Comment:
Nit: Is there any reason for using Long instead of long?

Reply:
Yes, the `intValue()` method.